src/loader/tsbgw_counter.c

```diff
@@ -0,0 +1,101 @@
+/* needed for initializing shared memory and using various locks */
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <storage/lwlock.h>
+#include <utils/hsearch.h>
+#include <storage/spin.h>
+#include <storage/shmem.h>
+#include <storage/ipc.h>
+#include <storage/latch.h>
+
+#include "tsbgw_counter.h"
+
+#define TSBGW_COUNTER_STATE_NAME "tsbgw_counter_state"
+/*
+ * We need a bit of shared state here to deal with keeping track of
+ * the total number of background workers we've launched across the instance
+ * since we don't want to exceed some configured value.
+ * We were excited, briefly, about the possibility of using pg_sema for this,
+ * unfortunately it does not appear to be accessible to code outside of postgres
+ * core in any meaningful way. So we're not using that.
+ */
+
+
+typedef struct tsbgwCounterState
+{
+	slock_t		mutex;			/* controls modification of total_workers */
+	int			total_workers;
+}			tsbgwCounterState;
+
+
+static tsbgwCounterState * tsbgw_counter_state_get(bool possible_restart){
+	static tsbgwCounterState * tsbgw_ct = NULL;
+	bool		found;
+
+	/* reset in case this is a restart within the postmaster */
+	if (possible_restart)
+		tsbgw_ct = NULL;
```

Comment:
Possible restart is set to true only when this is called in the shmem startup hook. The ShmemInitStruct is called either when initializing the struct or to find/map the struct in the backend. The only indication of whether it has been found is the found boolean and so we have to take the lock and be ready to initialize it whenever we call this, as far as I understand. Because it is only called with the possible restart boolean set in the shmem startup hook, we   don't do the heavyweight work anytime but then. The other option would be to keep a global var. The only reason we need the possible restart is that the Postmaster maps shared memory and so, when we have to do a hard reset of shared memory, it calls the shmem init hook in itself and needs to re-map shared mem because something might have been corrupted, so that's why we reset to NULL there and that should also be the only time we actually don't find it and initialize. Not sure if that helps explain? What of this should be in a comment in the code for future reference?

Reply:
Yes- in every backend, we will be called once when a) possible_restart (and I will switch the name) is true and b) tsbgw_ct is NULL, this will be the first call in every backend. So why do we need the possible restart variable: because there is one backend, specifically the postmaster, that may need to reinitialize shared memory in the case where another backend fails and possibly corrupts shared memory. At that point, tsbgw_ct will be non-null but we need to reinitialize because our current shared memory layout may have been corrupted. 
So on the first call in the postmaster we actually init the shared memory struct, on the first call to any other backend we simply map the struct (though the access method is the same, it just returns true for found) and assign the static variable. Then on subsequent calls the static variable is already set. Within the postmaster, if we have a restart, we run the shmem startup hook for a second time and need to reset the variable.

Reply:
Anyway. I think you're right, we should just replace with a global, so I'll change that pattern.